chrome/content/zotero/xpcom/data/feed.js

```diff
@@ -267,7 +295,150 @@ Zotero.Feed.prototype._finalizeSave = Zotero.Promise.coroutine(function* (env) {
 	}
 });
 
-Zotero.Feed.prototype._finalizeErase = Zotero.Promise.method(function(env) {
+Zotero.Feed.prototype.getExpiredFeedItemIDs = Zotero.Promise.coroutine(function* () {
+	let sql = "SELECT itemID AS id FROM feedItems "
+		+ "LEFT JOIN items I USING (itemID) "
+		+ "WHERE I.libraryID=? "
+		+ "AND readTime IS NOT NULL "
+		+ "AND julianday('now', 'utc') - (julianday(readTime, 'utc') + ?) > 0";
+	let expiredIDs = yield Zotero.DB.queryAsync(sql, [this.id, {int: this.cleanupAfter}]);
+	return expiredIDs.map(row => row.id);
+});
+
+Zotero.Feed.prototype.clearExpiredItems = Zotero.Promise.coroutine(function* () {
+	try {
+		// Clear expired items
+		if (this.cleanupAfter) {
+			let expiredItems = yield this.getExpiredFeedItemIDs();
+			Zotero.debug("Cleaning up read feed items...");
+			if (expiredItems.length) {
+				Zotero.debug(expiredItems.join(', '));
+				yield Zotero.FeedItems.forceErase(expiredItems);
+			} else {
+				Zotero.debug("No expired feed items");
+			}
+		}
+	} catch(e) {
+		Zotero.debug("Error clearing expired feed items.");
+		Zotero.debug(e);
+	}
+});
+
+Zotero.Feed.prototype._updateFeed = Zotero.Promise.coroutine(function* () {
+	this.updating = true;
```

Comment:
We should actually make sure that this doesn't get re-triggered until updating is done. Save a deferred promise in `this._updating` and resolve it at the end, then clear `this._updating`. Then, before anything else, return `this._updating` here if we're in the process of updating.